apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright 2022 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Http;
+using Google.Apis.Storage.v1;
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Storage.V1
+{
+    /// <summary>
+    /// Handler for exponential backoff for request messages which permit retries.
+    /// This is similar to BackoffHandler in Google.Apis.Http, but that doesn't give us enough flexibility.
+    /// This handler does not enforce a maximum number of retries; that's configured via <see cref="ConfigurableMessageHandler.NumTries"/>.
+    /// </summary>
+    internal sealed class RetryHandler : IHttpUnsuccessfulResponseHandler
+    {
+        private static readonly int[] s_retriableErrorCodes = { 408, 429, 500, 502, 503, 504 };
+        private static RetryHandler s_instance = new RetryHandler();
+
+        private RetryHandler() { }
+
+        internal static void MarkAsRetriable<TResponse>(StorageBaseServiceRequest<TResponse> request) =>
+            request.AddUnsuccessfulResponseHandler(s_instance);
+
+        //This function is designed to be asynchronous in case we need to examine the response content, but for now we only need the status code
+        internal static Task<bool> IsRetriableResponse(HttpResponseMessage response)
+        {
+            return Task.FromResult(s_retriableErrorCodes.Contains(((int)response.StatusCode)));
+        }
+
+        public async Task<bool> HandleResponseAsync(HandleUnsuccessfulResponseArgs args)
+        {
+            var retry = args.SupportsRetry && await IsRetriableResponse(args.Response).ConfigureAwait(false);
+            if (!retry)
+            {
+                return false;
+            }
+            // The first failure will have args.CurrentFailedTry set to 1,
+            // whereas we want the first delay to be 1 second. We use Math.Min on the power
+            // rather than on the result to obtain a max retry of 32 seconds without risking
+            // calling Math.Pow with a huge number.
+            int power = Math.Min(args.CurrentFailedTry - 1, 5);
```

Comment:
Are there currently retried RPCs in any of the dotnet libraries? If so is it possible to share the backoff timing code with them? Probably don't need a one-off for storage. In Go for example we use gax across products.  https://pkg.go.dev/github.com/googleapis/gax-go/v2

Reply:
We do have back-off code in GAX, but it's not easy to integrate into this particular code, where the only state we have is the `CurrentFailedTry`. (GAX code tends to assume that the retry sequence can be maintained between calls.)

I suspect it's simpler to stick to this effective duplication but simple integration than to try to somehow use the GAX code.